internal/expr/query.go

```diff
@@ -161,53 +151,25 @@ func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) (scanArgs []an
 		}
 		columnInResult[idx] = true
 		typeMember := qe.outputs[idx]
-		outputVal, ok := typeDest[typeMember.outerType()]
+		outputVal, ok := typeDest[typeMember.OuterType()]
 		if !ok {
-			return nil, nil, fmt.Errorf("type %q found in query but not passed to get", typeMember.outerType().Name())
+			return nil, nil, fmt.Errorf("type %q found in query but not passed to get", typeMember.OuterType().Name())
 		}
-		switch tm := typeMember.(type) {
-		case *structField:
-			val := outputVal.Field(tm.index)
-			if !val.CanSet() {
-				return nil, nil, fmt.Errorf("internal error: cannot set field %s of struct %s", tm.name, tm.structType.Name())
-			}
-			pt := reflect.PointerTo(val.Type())
-			if val.Type().Kind() != reflect.Pointer && !pt.Implements(scannerInterface) {
-				// Rows.Scan will return an error if it tries to scan NULL into a type that cannot be set to nil.
-				// For types that are not a pointer and do not implement sql.Scanner a pointer to them is generated
-				// and passed to Rows.Scan. If Scan has set this pointer to nil the value is zeroed.
-				scanVal := reflect.New(pt).Elem()
-				ptrs = append(ptrs, scanVal.Addr().Interface())
-				scanProxies = append(scanProxies, scanProxy{original: val, scan: scanVal})
-			} else {
-				ptrs = append(ptrs, val.Addr().Interface())
-			}
-		case *mapKey:
-			scanVal := reflect.New(tm.mapType.Elem()).Elem()
-			ptrs = append(ptrs, scanVal.Addr().Interface())
-			scanProxies = append(scanProxies, scanProxy{original: outputVal, scan: scanVal, key: reflect.ValueOf(tm.name)})
+
+		if ptrs, scanProxies, err = typeMember.AddScanTarget(outputVal, ptrs, scanProxies); err != nil {
```

Comment:
Here is the other.

Reply:
I completely agree with @Aflynn50, I think the code is clearer if we process then accumulate instead of mutating the state inside each function. Conceptually it is also not the role of a typeMember to mutate the "global" state that is used in this outer function, we are imposing state on a method that is really stateless.